File: src/main/java/com/github/nsc/de/shake/parser/node/CastNode.java
package com.github.nsc.de.shake.parser.node;

import com.github.nsc.de.shake.lexer.characterinput.position.PositionMap;

public class CastNode extends ValuedNode {

    private final ValuedNode value;
    private final CastTarget castTarget;

    public CastNode(PositionMap map, ValuedNode value, CastTarget castTarget) {
        super(map);
        this.value = value;
        this.castTarget = castTarget;
    }

    public ValuedNode getValue() {
        return value;
    }

    public CastTarget getCastTarget() {
        return castTarget;
    }

    public static class CastTarget {

        public static final CastTarget BYTE = new CastTarget(CastTargetType.BYTE);
        public static final CastTarget SHORT = new CastTarget(CastTargetType.SHORT);
        public static final CastTarget INTEGER = new CastTarget(CastTargetType.INTEGER);
        public static final CastTarget LONG = new CastTarget(CastTargetType.LONG);
        public static final CastTarget FLOAT = new CastTarget(CastTargetType.FLOAT);
        public static final CastTarget DOUBLE = new CastTarget(CastTargetType.DOUBLE);
        public static final CastTarget BOOLEAN = new CastTarget(CastTargetType.BOOLEAN);
        public static final CastTarget CHAR = new CastTarget(CastTargetType.CHAR);
        public static final CastTarget STRING = new CastTarget(CastTargetType.STRING);

        private final CastTargetType type;
        private final IdentifierNode subtype;

        public CastTarget(CastTargetType type, IdentifierNode subtype) {
            this.type = type;
            this.subtype = subtype;
        }

        public CastTarget(CastTargetType type) {
            this(type, null);
        }

        public CastTarget(IdentifierNode type) {
            this(CastTargetType.OBJECT, type);
        }

        public CastTargetType getType() {
            return type;
        }

        public IdentifierNode getSubtype() {
            return subtype;
        }

        public enum CastTargetType {
            BYTE, SHORT, INTEGER, LONG, FLOAT, DOUBLE, BOOLEAN, CHAR, STRING, OBJECT
        }

    }

}

File: parser/src/commonMain/kotlin/io/github/shakelang/shake/parser/node/CastNode.kt
package io.github.shakelang.shake.parser.node

import io.github.shakelang.parseutils.characters.position.PositionMap
import kotlin.jvm.JvmField
import kotlin.jvm.JvmOverloads

@Suppress("unused")
class CastNode(map: PositionMap, val value: ValuedNode, val castTarget: CastTarget) : ValuedNode(map) {

    class CastTarget @JvmOverloads constructor(val type: CastTargetType, val subtype: IdentifierNode? = null) {

        constructor(type: IdentifierNode?) : this(CastTargetType.OBJECT, type)

        enum class CastTargetType {
            BYTE, SHORT, INT, LONG, FLOAT, DOUBLE, BOOLEAN, CHAR, STRING, OBJECT
        }

        override fun toString(): String {
            return subtype?.toString() ?: type.toString()
        }

        companion object {
            @JvmField
            val BYTE = CastTarget(CastTargetType.BYTE)
            @JvmField
            val SHORT = CastTarget(CastTargetType.SHORT)
            @JvmField
            val INTEGER = CastTarget(CastTargetType.INT)
            @JvmField
            val LONG = CastTarget(CastTargetType.LONG)
            @JvmField
            val FLOAT = CastTarget(CastTargetType.FLOAT)
            @JvmField
            val DOUBLE = CastTarget(CastTargetType.DOUBLE)
            @JvmField
            val BOOLEAN = CastTarget(CastTargetType.BOOLEAN)
            @JvmField
            val CHAR = CastTarget(CastTargetType.CHAR)
            @JvmField
            val STRING = CastTarget(CastTargetType.STRING)
        }
    }

    override fun toJson(): Map<String, *> =
        mapOf(
            "name" to "CastNode",
            "value" to value.json,
            "cast_target" to castTarget.toString()
        )
}